src/parser/jobs/rdm/Procs.js

```diff
@@ -0,0 +1,322 @@
+import React, {Fragment} from 'react'
+import {ActionLink, StatusLink} from 'components/ui/DbLink'
+import ACTIONS from 'data/ACTIONS'
+import STATUSES from 'data/STATUSES'
+import Module from 'parser/core/Module'
+import {TieredSuggestion} from 'parser/core/modules/Suggestions'
+import {CAST_STATE, PROCS, SEVERITY_MISSED_PROCS, SEVERITY_OVERWRITTEN_PROCS, SEVERITY_INVULN_PROCS} from 'parser/jobs/rdm/ProcsEnum'
+import {i18nMark} from '@lingui/react'
+
+export default class Procs extends Module {
+	static handle = 'procs'
+	static title = 'Procs'
+	static dependencies = [
+		'aoe',
+		'castTime',
+		'downtime',
+		'gcd',
+		'suggestions',
+		'gauge',
+	]
+	static i18n_id = i18nMark('rdm.procs.title')
+
+	_history = {
+		wastedStone: 0,
+		overWrittenStone: 0,
+		invulnStone: 0,
+		missedStone: 0,
+		wastedFire: 0,
+		overWrittenFire: 0,
+		invulnFire: 0,
+		missedFire: 0,
+		wastedImpactful: 0,
+		overWrittenImpactful: 0,
+		invulnImpactful: 0,
+		missedImpactful: 0,
+		wastedEnhancedScatters: 0,
+		invulnScatter: 0,
+		missedScatter: 0,
+	}
+	_currentProcs = {
+		verStoneGained: 0,
+		verFireGained: 0,
+		impactfulGained: 0,
+		enhancedScatterGained: 0,
+	}
+	_castState = CAST_STATE.NA
+	_impactfulProcOverride = false
+	_previousCast = null
+	_bossWasInvuln = false
+
+	constructor(...args) {
+		super(...args)
+
+		this.addHook('cast', {by: 'player'}, this._onCast)
+		this.addHook('applybuff', {
+			to: 'player',
+			abilityId: PROCS,
+		}, this._onGain)
+		this.addHook('removebuff', {
+			to: 'player',
+			abilityId: PROCS,
+		}, this._onRemove)
+		this.addHook('complete', this._onComplete)
+	}
+
+	_onCast(event) {
+		const abilityID = event.ability.guid
+		const invuln = this.downtime.getDowntime(this._previousCast||0, event.timestamp)
+		this._previousCast = event.timestamp
+		this._bossWasInvuln = invuln > 0
+
+		switch (abilityID) {
```

Comment:
I think this could be hugely simplified, by just going to use the ACTION id instead of this custom enum.
I don't see any reason for the enum to be there quite honestly.

To simplify this, you could make a bunch of fall through switch cases, then assign the cast state being the casts id, defaulting to setting it to null and making special treatment to impact.

Reply:
Note - no language I've worked on prior has allowed fall through switch cases, so I'm not sure what it's functional application is.

In this particular instance, I'm not certain what you would fall through to though, each case beyond impactful is only setting the current cast state so I can accurately determine what was cast immediately prior to a buff being removed, the buff removing happens in a different event so I wouldn't be able to query the onRemove event for the spell that was cast to trigger the removal.  I could remove the enum, I prefer working with subsets, but not opposed to working with the full set.  Or are you stating I should just ditch the State capture and capture the previously cast ability ID and then query it from ACTIONS in my other functions?

Reply:
Something like this, although honestly I'd just not do a switch statement and make some logic to ignore other `_castState` values and make the impact stuff just an if block.

```javascript
switch(actionId) {
case ACTIONS.IMPACT.id:
    this._castState = actionId
    // [...] Impact specific stuff that you already have
    break
case ACTIONS.VERSTONE.id:
case ACTIONS.VERFIRE.id:
case ACTIONS.VERTHUNDER.id:
case ACTIONS.VERAREO.id: // is this a typo in the ACTIONS map?
case ACTIONS.SCATTER.id:
    this._castState = actionId
    break
default:
    this._castState = null
    break
}
```